numpy/core/arrayprint.pyi

```diff
@@ -0,0 +1,148 @@
+import sys
+from types import TracebackType
+from typing import Any, Optional, Callable, Union, Type
+
+# Using a private class is by no means ideal, but it is simply a consquence
+# of a `contextlib.context` returning an instance of aformentioned class
+from contextlib import _GeneratorContextManager
+
+from numpy import (
+    ndarray,
+    generic,
+    bool_,
+    integer,
+    timedelta64,
+    datetime64,
+    floating,
+    complexfloating,
+    void,
+    str_,
+    bytes_,
+    longdouble,
+    clongdouble,
+)
+from numpy.typing import ArrayLike, _CharLike, _FloatLike
+
+if sys.version_info > (3, 8):
+    from typing import Literal, TypedDict
+else:
+    from typing_extensions import Literal, TypedDict
+
+class _FormatDict(TypedDict, total=False):
+    bool: Callable[[bool_], str]
+    int: Callable[[integer[Any]], str]
+    timedelta: Callable[[timedelta64], str]
+    datetime: Callable[[datetime64], str]
+    float: Callable[[floating[Any]], str]
+    longfloat: Callable[[longdouble], str]
+    complexfloat: Callable[[complexfloating[Any, Any]], str]
+    longcomplexfloat: Callable[[clongdouble], str]
+    void: Callable[[void], str]
+    numpystr: Callable[[_CharLike], str]
+    object: Callable[[object], str]
+    str: Callable[[Any], str]  # Unused but still present?
+    all: Callable[[object], str]
+    int_kind: Callable[[integer[Any]], str]
+    float_kind: Callable[[floating[Any]], str]
+    complex_kind: Callable[[complexfloating[Any, Any]], str]
+    str_kind: Callable[[_CharLike], str]
+
+class _FormatOptions(TypedDict):
+    precision: int
+    threshold: int
+    edgeitems: int
+    linewidth: int
+    suppress: bool
+    nanstr: str
+    infstr: str
+    formatter: Optional[_FormatDict]
+    sign: Literal["-", "+", " "]
+    floatmode: Literal["fixed", "unique", "maxprec", "maxprec_equal"]
+    legacy: Literal[False, "1.13"]
+
+def set_printoptions(
+    precision: Optional[int] = ...,
+    threshold: Optional[int] = ...,
+    edgeitems: Optional[int] = ...,
+    linewidth: Optional[int] = ...,
+    suppress: Optional[bool] = ...,
+    nanstr: Optional[str] = ...,
+    infstr: Optional[str] = ...,
+    formatter: Optional[_FormatDict] = ...,
+    sign: Optional[Literal["-", "+", " "]] = ...,
+    floatmode: Optional[
+        Literal["fixed", "unique", "maxprec", "maxprec_equal"]
```

Comment:
This might be worth abstracting to a `_FloatMode =  Literal["fixed", "unique", "maxprec", "maxprec_equal"]`

Reply:
Not a bad idea, done as of https://github.com/numpy/numpy/pull/18032/commits/05fa6e4bc219fb4db0808ef5cf9aed96c8ec0844.